### PR TITLE
Try evaluating constants eagerly

### PR DESCRIPTION
**Pull request by tmiasko**

### CODE REVIEW

**Comment by matthewjasper** [comment]
cc @rust-lang/wg-mir-opt

File: src/test/mir-opt/remove-never-const.rs [hunk 1]
```diff
@@ -1,23 +0,0 @@
...
-#![feature(never_type)]
-#![warn(const_err)]
-
-struct PrintName<T>(T);
-
-impl<T> PrintName<T> {
-    const VOID: ! = panic!();
-}
-
-// EMIT_MIR remove_never_const.no_codegen.PreCodegen.after.mir
```
Comment (oli-obk) [inline src/test/mir-opt/remove-never-const.rs:18]: why was this test removed? Does it error now?

Comment (tmiasko) [reply src/test/mir-opt/remove-never-const.rs:18]: When compiling with `--emit mir` it now fails with:

```
error[E0080]: any use of this value will cause an error
  --> src/test/mir-opt/remove-never-const.rs:15:21
   |
15 |     const VOID: ! = panic!();
   |     ----------------^^^^^^^^-
   |                     |
   |                     the evaluated program panicked at 'explicit panic', src/test/mir-opt/remove-never-const.rs:15:21
```

**Comment by tmiasko** [comment]
This is large improvement for encoding-debug benchmark (-15.1% full, -18.2% incr-full). This crate uses a number of inline functions from other crates. Those functions in turn reference large constant arrays which are now pre-evaluated. This also suggest an idea for a lint that detects large constants that could be turned into statics.

The impact on check builds is mostly negligible with exception of larger 1.9% regression in deeply-nested-async.

The debug and opt builds are mix of improvements and regression in -0.7%..+0.7% range.

**Pull request closed without merging**